src/components/app.cjsx

```diff
@@ -1,35 +1,140 @@
 React = require 'react'
 BS = require 'react-bootstrap'
+classnames = require 'classnames'
 Exercise = require './exercise'
+ErrorModal = require './error-modal'
+{ExerciseActions, ExerciseStore} = require '../stores/exercise'
+AsyncButton = require 'openstax-react-components/src/components/buttons/async-button.cjsx'
 
 module.exports = React.createClass
   displayName: 'App'
-  getInitialState: -> {exerciseId: 'new'}
+  getInitialState: -> {verticalDisplay: true}
 
-  update: (event) ->
-    exerciseId = event.target.value
+  update: -> @forceUpdate()
+
+  componentWillMount: ->
+    ExerciseStore.addChangeListener(@update)
+
+  componentWillUnmount: ->
+    ExerciseStore.removeChangeListener(@update)
+
+  componentDidMount: ->
+    if @props.exerciseId is 'new'
+      @addNew()
+    else if @props.exerciseId
+      this.refs.exerciseId.getDOMNode().value = @props.exerciseId
+      @loadExercise()
+
+  publishExercise: ->
+    if confirm('Are you sure you want to publish?')
+      ExerciseActions.save(@state.exerciseId)
+      ExerciseActions.publish(@state.exerciseId)
+
+  redirect:(id) ->
+    window.location = "/exercises/#{id}"
+
+  addNew: ->
+    id = ExerciseStore.freshLocalId()
+    template = ExerciseStore.getTemplate()
+    ExerciseActions.loaded(template, id)
+    @setState({exerciseId: id })
+
+  loadExercise: ->
+    exerciseId = this.refs.exerciseId.getDOMNode().value
     @setState({exerciseId})
+    ExerciseActions.load(exerciseId)
+
+  onFindKeyPress: (ev) ->
+    return unless ev.key is 'Enter'
+    @loadExercise()
+    ev.preventDefault()
+
+  toggleVerticalDisplay: (ev) ->
+    @setState(verticalDisplay: ev.target.checked)
+
+  renderExerciseOrLoad: ->
+    if @state.exerciseId and not ExerciseStore.isNew(@state.exerciseId)
+      <div>Exercise ID: {@state.exerciseId}</div>
+    else
+      <div className="input-group">
+        <input type="text" className="form-control" onKeyPress={@onFindKeyPress}
+          ref="exerciseId" placeholder="Exercise ID"/>
+        <span className="input-group-btn">
+          <button className="btn btn-default" type="button" onClick={@loadExercise}>Load</button>
+        </span>
+      </div>
+
+  renderEditingBody: ->
+    return null unless @state.exerciseId
+    if ExerciseStore.isLoading()
+      <div>Loading...</div>
```

Comment:
can you add classes similar to `<Loadable>` so selenium tests would work (even though we don't have them for exercises yet)?